src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/Routing/ActionEndpointFactory.cs

```diff
@@ -0,0 +1,278 @@
+using System.Linq;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Abstractions;
+using Microsoft.AspNetCore.Mvc.ActionConstraints;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Patterns;
+
+namespace Microsoft.AspNetCore.Mvc.Routing
+{
+    internal class ActionEndpointFactory
+    {
+        private readonly RoutePatternTransformer _routePatternTransformer;
+        private readonly MvcEndpointInvokerFactory _invokerFactory;
+
+        public ActionEndpointFactory(
+            RoutePatternTransformer routePatternTransformer,
+            MvcEndpointInvokerFactory invokerFactory)
+        {
+            if (routePatternTransformer == null)
+            {
+                throw new ArgumentNullException(nameof(routePatternTransformer));
+            }
+
+            if (invokerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(invokerFactory));
+            }
+
+            _routePatternTransformer = routePatternTransformer;
+            _invokerFactory = invokerFactory;
+        }
+
+        public void AddEndpoints(
+            List<Endpoint> endpoints,
+            ActionDescriptor action,
+            IReadOnlyList<ConventionalRouteEntry> routes,
+            IReadOnlyList<Action<EndpointBuilder>> conventions)
+        {
+            if (endpoints == null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            if (action == null)
+            {
+                throw new ArgumentNullException(nameof(action));
+            }
+
+            if (routes == null)
+            {
+                throw new ArgumentNullException(nameof(routes));
+            }
+
+            if (conventions == null)
+            {
+                throw new ArgumentNullException(nameof(conventions));
+            }
+
+            if (action.AttributeRouteInfo == null)
+            {
+                // In traditional conventional routing setup, the routes defined by a user have a static order
+                // defined by how they are added into the list. We would like to maintain the same order when building
+                // up the endpoints too.
+                //
+                // Start with an order of '1' for conventional routes as attribute routes have a default order of '0'.
+                // This is for scenarios dealing with migrating existing Router based code to Endpoint Routing world.
+                var conventionalRouteOrder = 1;
+
+                // Check each of the conventional patterns to see if the action would be reachable.
+                // If the action and pattern are compatible then create an endpoint with action
+                // route values on the pattern.
+                foreach (var route in routes)
+                {
+                    // An 'endpointInfo' is applicable if:
```

Comment:
What does endpointInfo refer to? I'm guessing the comment is out of date

Reply:
Yep. will fix!